powertools-parameters/src/main/java/software/amazon/lambda/powertools/parameters/BaseProvider.java

```diff
@@ -0,0 +1,186 @@
+     */
+    protected BaseProvider withTransformation(Class<? extends Transformer> transformerClass) {
+        if (transformationManager == null) {
+            throw new IllegalStateException("Trying to add transformation while no TransformationManager has been provided.");
+        }
+        transformationManager.setTransformer(transformerClass);
+        return this;
+    }
+
+    /**
+     * Get the value of a parameter, either from the underlying store or a cached value (if not expired).<br/>
+     * Using this method, you can apply a basic transformation (to String). <br/>
+     * Set a {@link BasicTransformer} with {@link #withTransformation(Class)}.<br/><br/>
+     * If you need a more complex transformation (to Object), use {@link #get(String, Class)} method instead of this one. <br/>
+     *
+     * @param key key of the parameter
+     * @return the String value of the parameter
+     * @throws IllegalArgumentException if a wrong transformer class is provided through {@link #withTransformation(Class)}. Needs to be a {@link BasicTransformer}.
+     * @throws TransformationException  if the transformation could not be done, because of a wrong format or an error during transformation.
+     */
+    @Override
+    public String get(String key) {
+        try {
+            Optional<String> valueInCache = cacheManager.getIfNotExpired(key);
+            if (valueInCache.isPresent()) {
+                return valueInCache.get();
+            }
+
+            String value = getValue(key);
+
+            String transformedValue = value;
+            if (transformationManager != null && transformationManager.shouldTransform()) {
+                transformedValue = transformationManager.performBasicTransformation(value);
+            }
+
+            cacheManager.putInCache(key, transformedValue);
+
+            return transformedValue;
+
+        } finally {
+            // in all case, we reset options to default, for next call
+            resetToDefaults();
+        }
+    }
+
+    /**
+     * Get the value of a parameter, either from the underlying store or a cached value (if not expired).<br/>
+     * Using this method, you must apply a transformation (eg. json/xml to Object). <br/>
+     * Set a {@link Transformer} with {@link #withTransformation(Class)}.<br/><br/>
+     * If you need a simpler transformation (to String), use {@link #get(String)} method instead of this one.
+     *
+     * @param key         key of the parameter
+     * @param targetClass class of the target Object (after transformation)
+     * @return the Object (T) value of the parameter
+     * @throws IllegalArgumentException if no transformation class was provided through {@link #withTransformation(Class)}
+     * @throws TransformationException  if the transformation could not be done, because of a wrong format or an error during transformation.
+     */
+    @Override
+    public <T> T get(String key, Class<T> targetClass) {
```

Comment:
Minor: May be we can wrap try finally logic like this?

```
  private <R> R withResetToDefaults(String key, Function<String, R> function) {
        try {
            return function.apply(key);
        } finally {
            resetToDefaults();
        }
    }

```

And then both the method could just call this wrapper ?

Reply:
2 different signatures, don't think it's worth it.

Reply:
Ok no problem.